# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit-test/tests/debug/Frame-this-15.js
# Commit: 31d14cf1975c
# Full Hash: 31d14cf1975c1b23ef3db176252f8cc3b31a0772
# Author: Logan Smyth <loganfsmyth@gmail.com>
# Date: 2019-12-09 09:50:39
# Regressor Bug: 1600204
# File Overlap Count: 2
# Description:
#   Bug 1600204 - Part 7: Support .this on suspended generator frames. r=jimb
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D55580
# ==============================================================================

diff -r fa46d8caefde -r 31d14cf1975c js/src/jit-test/tests/debug/Frame-this-15.js
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/js/src/jit-test/tests/debug/Frame-this-15.js	Sun Dec 08 23:54:38 2019 +0000
@@ -0,0 +1,34 @@
+// Test that Debugger.Frame.prototype.this works on a suspended async
+// function.
+
+load(libdir + "asserts.js");
+
+const g = newGlobal({ newCompartment: true });
+const dbg = new Debugger();
+const gDO = dbg.addDebuggee(g);
+
+g.eval(`
+var context = {};
+var f = async function() {
+  await Promise.resolve();
+  return this;
+}.bind(context);
+`);
+
+let frame;
+dbg.onEnterFrame = f => {
+  frame = f;
+  assertEq(frame.this, gDO.makeDebuggeeValue(g.context));
+  dbg.onEnterFrame = undefined;
+};
+
+(async () => {
+  const promise = g.f();
+
+  assertEq(!!frame, true);
+  assertEq(frame.this, gDO.makeDebuggeeValue(g.context));
+
+  await promise;
+
+  assertThrowsInstanceOf(() => frame.this, Error);
+})();